module-system/module-implementations/sov-bank/src/hooks.rs

```diff
@@ -0,0 +1,67 @@
+use std::sync::Arc;
+
+use sov_modules_api::hooks::TxHooks;
+use sov_modules_api::transaction::Transaction;
+use sov_modules_api::{Context, WorkingSet};
+
+use crate::{Bank, Coins};
+
+/// The computed addresses to deduct the gas from.
+pub struct BankTxHook<C: Context> {
+    /// The tx sender address
+    pub sender: Arc<C::Address>,
+    /// The sequencer address
+    pub sequencer: Arc<C::Address>,
+}
+
+impl<C: Context> TxHooks for Bank<C> {
+    type Context = C;
+    type PreArg = BankTxHook<C>;
+    type PreResult = ();
+
+    fn pre_dispatch_tx_hook(
+        &self,
+        tx: &Transaction<C>,
+        working_set: &mut WorkingSet<C>,
+        arg: BankTxHook<C>,
+    ) -> anyhow::Result<()> {
+        let BankTxHook { sender, sequencer } = arg;
+
+        let amount = tx.gas_limit().saturating_add(tx.gas_tip());
+        if amount > 0 {
+            let token_address = self.address.clone();
```

Comment:
This is the address of the module, but we have to transfer from the address of the gas `token`.

Reply:
Done!